br/pkg/lightning/restore/data_sample_check.go

```diff
@@ -0,0 +1,254 @@
+		}
+		// see Datum.String()
+		v := d.GetValue()
+		if b, ok := v.([]byte); ok && d.Kind() == types.KindBytes {
+			v = string(b)
+		}
+		sb.WriteString(fmt.Sprintf("%v", v))
+	}
+	log.L().Error("data check failed", zap.String("raw-data", sb.String()),
+		zap.String("file", fileMeta.Path),
+		zap.Int64("offset", offset),
+		zap.String("error", errMsg))
+}
+
+func (d *dataSampleCheck) checkRoutine(ctx context.Context, fileChan chan *sampledDataFileInfo, errChan chan error) {
+	defer d.wg.Done()
+
+	var resultErr error
+	defer func() {
+		errChan <- errors.Trace(resultErr)
+	}()
+
+	rc := d.controller
+	for fileInfo := range fileChan {
+		fileMeta := fileInfo.DataFile.FileMeta
+		fileParser, err := rc.createDataFileParser(ctx, fileMeta, config.DefaultCSVDataInvalidCharReplace)
+		if err != nil {
+			if errors.Cause(err) != context.Canceled {
+				resultErr = err
+			}
+			return
+		}
+
+		var rowsChecked, columnCountMismatchRows, invalidCharRows, lastOffset int64
+		columnCount := len(fileInfo.TableInfo.Columns)
+		for i := 0; i < d.checkCfg.Rows; i++ {
+			err = fileParser.ReadRow()
+			if err != nil {
+				if errors.Cause(err) == io.EOF {
+					break
+				}
+
+				if errors.Cause(err) != context.Canceled {
+					resultErr = err
+				}
+				fileParser.Close()
+				return
+			}
+
+			row := fileParser.LastRow()
+			rowsChecked++
+			if len(row.Row) != columnCount {
+				columnCountMismatchRows++
+				d.logCheckFailedRow(row.Row, fileMeta, lastOffset,
+					fmt.Sprintf("column count mismatch, expect %d, got %d", columnCount, len(row.Row)))
+			}
+
+			for _, col := range row.Row {
+				// for non-string column, it returns empty string
+				colStr := col.GetString()
+				if !utf8.ValidString(colStr) || strings.Contains(colStr, config.DefaultCSVDataInvalidCharReplace) {
+					invalidCharRows++
+					d.logCheckFailedRow(row.Row, fileMeta, lastOffset, "contains invalid char")
+				}
+			}
+			fileParser.RecycleRow(row)
+			// lastOffset is the start offset of current row
+			lastOffset, _ = fileParser.Pos()
+		}
+
+		fileParser.Close()
+
+		d.totalRows.Add(rowsChecked)
+		d.totalColumnCountMismatchRows.Add(columnCountMismatchRows)
+		d.totalInvalidCharRows.Add(invalidCharRows)
+	}
+}
+
+func (d *dataSampleCheck) doCheck(ctx context.Context) error {
+	rc := d.controller
+
+	targetFiles, err := d.getSampledDataFiles()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	concurrency := utils.MinInt(runtime.NumCPU(), 8)
+	fileChan := make(chan *sampledDataFileInfo, concurrency)
+	errChan := make(chan error, concurrency)
+	newCtx, cancelFunc := context.WithCancel(ctx)
+
+	for i := 0; i < concurrency; i++ {
+		d.wg.Add(1)
+		go d.checkRoutine(newCtx, fileChan, errChan)
+	}
+
+	for _, fileInfo := range targetFiles {
+		fileChan <- fileInfo
+	}
+	close(fileChan)
+
+	// we need to exit if any error occurs in check routine
+	// each routine should return a result in the error channel, return nil if there's no error.
+	for i := 0; i < concurrency; i++ {
```

Comment:
Since we only check at most one error, I think replace the errChan with common.Once is more ergonimic. In this way, we can just wait d.wg.Wait and then check if there is error occurs.

Reply:
we need to cancel other routines without error